### ISSUE TITLE
Inverted logic for smart gaps

### ISSUE THREAD

**Issue by 9ary** [Airblader/i3#263]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior
With the current smart_gaps option, gaps can only be hidden when there is _one_ window on the current workspace.

## Desired Behavior
It should be possible to invert that behavior, ie gaps with one window but no gaps with multiple windows. Affecting only inner or outer gaps should also be an option.
Use case: I have a 27" monitor, and a single window is simply too big most of the time in this configuration. Horizontal gaps help constrain that to a workable width, but the current logic is backwards for me.

**Comment by Airblader** [comment]
I would be OK with adding another value for `smart_gaps` that toggles the gaps, but also controlling different gaps individually is getting a bit out of too hand for my taste – it adds a lot of complexity, and I would encourage solving this with an IPC script instead.

However, this leaves the question whether the new value would even be of any use – as I understand it, the answer is no?

**Comment by 9ary** [comment]
> controlling different gaps individually is getting a bit out of too hand for my taste – it adds a lot of complexity, and I would encourage solving this with an IPC script instead.

I don't think controlling each edge separately would be necessary. For inner gaps, only the current implementation makes sense to me. This leaves a few combinations of the following:
- inner gaps can be smart or non-smart
- outer gaps can be smart, inverse-smart or non-smart

We could just go ahead and pick one or two in addition to the current modes (all smart or all non-smart). More complex use cases should indeed rely on IPC scripting.

> However, this leaves the question whether the new value would even be of any use – as I understand it, the answer is no?

I'm not sure I understand what you mean here.

**Comment by Airblader** [comment]
> This leaves a few combinations of the following: [...] We could just go ahead and pick one or two in addition to the current modes

Can you write up an actual proposal of the syntax options you're thinking of? As I said, I don't think. I'm willing to add all these different options beyond a simple "inverse smart", though, as it gets really feature creepy at this point and it becomes hard to argue why we have X and not Y. It's really not hard to script this anyway.

> I'm not sure I understand what you mean here.

Assuming we only add an inverse smart mode, but no further distinction, would this feature even still be useful to you?

**Comment by 9ary** [comment]
> Can you write up an actual proposal of the syntax options you're thinking of?

I'd say `smart_gaps inverse`, or `smart_gaps inverse_outer` if you consider my proposal below.

> Assuming we only add an inverse smart mode, but no further distinction, would this feature even still be useful to you?

I'd say yes. Ultimately, the goal is to avoid strain from constantly looking around, not to make it look pretty. I've been using upstream i3 and this is pretty much the only reason why I would use the gaps patch.

---
Tangent: let's see if I can find a combination that makes sense. Leaving out the two existing modes, and everything that involves "inverse-smart" inner gaps (like I said, it'd be weird to want inner gaps to only show with one window), we're left with this:

|inner|outer|comment|conclusion|
|-|-|-|-|
|smart|inv-smart|can be reproduced with non-smart+inv-smart and carefully chosen outer gaps (negative values are allowed)|no|
|smart|non-smart|same as above|no|
|non-smart|smart|I'm not quite sure what one would use this combination for, but it can't be matched with any other mode|maybe|
|non-smart|inv-smart|seems like the most flexible mode|yes|

All things considered I think the last option would be better than making all gaps inv-smart and should cover most use-cases, like you said anything else is easy to script.

**Comment by Airblader** [comment]
Sounds fair. Would you be willing to put together a PR for this so we can decide based on the complexity? I assume it shouldn't be too drastic of a change, but I'd like to see it first, so a prototype PR would make it easy to judge.

**Comment by 9ary** [comment]
Yeah I'll work on this today.

**Comment by apiep** [comment]
Is this issue resolved and the PR merged into the latest release? Because I still have an error when setting `smart_gaps` to `inverse_outer`

```
ERROR: CONFIG: Expected one of these tokens: '1', 'yes', 'true', 'on', 'enable', 'active'
ERROR: CONFIG: (in file /home/work/.config/i3/config)
ERROR: CONFIG: Line  24: gaps outer 200
ERROR: CONFIG: Line  25: # smart_gaps inverse_outer
ERROR: CONFIG: Line  26: smart_gaps inverse_outer
ERROR: CONFIG:                      ^^^^^^^^^^^^^
ERROR: CONFIG: Line  27: set --no-startup-id $TERMINAL --no-startup-id --no-startup-id $TERMINALINAL
ERROR: CONFIG: Line  28: set Mod4 Mod4
ERROR: FYI: You are using i3 version 4.16.1 (2019-01-27)
```

**Comment by Airblader** [comment]
@apiep This was merged in December, 4.16 is from November. 4.16.1 is from January, but it's a patch release and thus didn't include new features but only bugfixes.

In other words this feature will be in 4.17.

**Comment by Airblader** [comment]
… That said, this should be closed now ;-)
